Add dense_vector equality and inequality support in ES|QL

majiayu000 opened the pull request:
## Summary

Adds support for binary comparison operators (==, !=) for dense_vector type in ES|QL.

Resolves #139929

## Changes

- Added `EqualsDenseVectorEvaluator` and `NotEqualsDenseVectorEvaluator` classes to handle dense_vector comparison
- Updated `Equals.java` and `NotEquals.java` with dense_vector support in evaluatorMap and @Param annotations
- Added `DENSE_VECTOR_EQUALITY` capability in EsqlCapabilities
- Updated test classes with dense_vector test cases
- Added csv-spec tests for dense_vector equality operations

Two dense vectors are equal if they have the same dimensions and all corresponding elements are equal.

## Test plan

- [x] EqualsTests with dense_vector test cases
- [x] NotEqualsTests with dense_vector test cases
- [x] EqualsErrorTests updated
- [x] NotEqualsErrorTests updated
- [x] dense_vector.csv-spec integration tests

elasticsearchmachine commented:
Pinging @elastic/es-search-relevance (Team:Search Relevance)

carlosdelest commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/dense_vector.csv-spec:
```diff
@@ -105,3 +105,91 @@ row hex_str = "0102030405060708090a0b0c0d0e0f"
+required_capability: dense_vector_agg_metric_double_if_version
+
+FROM dense_vector
+| EVAL same = float_vector == float_vector
```
I think it would be better to test the actual data with a specific dense vector, like:

```
FROM dense_vector
| EVAL same = float_vector == [1.0, 2.0, 3.0]
```

carlosdelest commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/dense_vector.csv-spec:
Expanding to the other dense_vector element types would be great as well - the same tests should be added to `dense_vector-bit.csv-spec`, `dense_vector-byte.csv-spec`, `dense_vector-bfloat16.csv-spec` 👍

carlosdelest replied:
Nice, thanks!

carlosdelest reviewed (commented):
Nice work!

Some comments about adding test cases - can you please address them @majiayu000 ?

carlosdelest commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/dense_vector.csv-spec:
Thanks for addressing the comments!

Let's add some tests for different dimensions for equality and inequality so we test those cases out as well.

majiayu000 commented:
Hi @carlosdelest, thanks for the additional feedback! I have updated the CSV specs to:
1. Simplify the capabilities to only require `dense_vector_equality` as suggested.
2. Add test cases for dimension mismatch (e.g., `denseVectorCompareDifferentDimensions`) in all 4 spec files.
Let me know if this looks good!

carlosdelest commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/EqualsTests.java:
```diff
@@ -252,11 +252,42 @@ public static Iterable<Object[]> parameters() {
+                denseVectorCases,
+                denseVectorCases,
+                List.of(),
+                false
```
Symmetric should be true here IMO, as equals / not equals is symmetric:

```suggestion
                true
```

kkharbas commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/NotEqualsDenseVectorEvaluator.java:
```diff
@@ -0,0 +1,118 @@
+ * {@link EvalOperator.ExpressionEvaluator} implementation for comparing dense_vector inequality.
+ * Two dense vectors are considered not equal if they have different dimensions or any elements differ.
+ */
+public final class NotEqualsDenseVectorEvaluator implements EvalOperator.ExpressionEvaluator {
```
minor comment/question - in interest of code de-duplication, this evaluator can be written in terms of `EqualsDenseVectorEvaluator ` @carlosdelest have we used such pattern any where else?

carlosdelest commented:
@elasticsearchmachine test this please

majiayu000 commented:
@carlosdelest Thanks for the review!

I've fixed the compilation errors in EqualsTests and NotEqualsTests (missing static imports). Regarding the test coverage, you were right — the fixed set of random tests wasn't ideal. I've refactored it to use the TestCaseSupplier.casesCrossProduct approach with a stable set of random vectors. This ensures we test the full cross-product (both equality and inequality scenarios) while keeping the implementation clean.

Ready for another look!

majiayu000 committed:
Add dense_vector equality and inequality support in ES|QL

Signed-off-by: majiayu000 <1835304752@qq.com>

majiayu000 committed:
Refactor InsensitiveEqualsTests to use TestCaseSupplier

carlosdelest commented:
@majiayu000 sorry but those commits you mention are not part of any branch on the repo. [This](https://github.com/elastic/elasticsearch/pull/140005/changes#diff-633e999be3313d604885978e4fb0c59802fcf9f3bd12cacea563afe253f766ef) is what I see for `EqualsTests`, and I don't see any changes done there.

Please double check that you're pushing the right branch to the repo and the changes in this PR are what you expect. Thanks!

majiayu000 committed:
Trigger PR update

carlosdelest reviewed (commented):
Hey @majiayu000 , I added more context on what I expect to see on the tests.

Also, I keep having `precommit` errors. Please follow up the [formatting guidelines](https://github.com/elastic/elasticsearch/blob/cbe110a70c3038c5aacfc3e230753bcc990ab523/CONTRIBUTING.md#java-language-formatting-guidelines), and ensure that everything is OK by running `./gradlew :x-pack:plugin:esql:precommit --configuration-cache` before pushing.

Thanks!

majiayu000 committed:
Fix dense_vector equality tests per review comments

majiayu000 commented:
Thanks for handling the conflicts! I don’t have any further changes or pending work on my end. It’s ready for the next round of review. Thanks!

Commit:
[CI] Auto commit changes from spotless

kkharbas commented:
@majiayu000 thank you for your contribution. PR looks good % one comment - the updates to `InsensitiveEqualsTests` don't seem to be related to dense_vector equality.

majiayu000 committed:
Revert InsensitiveEqualsTests changes unrelated to dense_vector equality

majiayu000 commented:
@kkharbas Thanks for the review\! You're right — the `InsensitiveEqualsTests` changes were unrelated. I've reverted them in the latest commit.

mromaios committed:
refactor: use existing testing constructs in Equals and NotEquals tests

mromaios commented:
@elasticsearchmachine test this please

majiayu000 committed:
Fix dense_vector csv-spec casts and null expectations

Signed-off-by: majiayu000 <1835304752@qq.com>

mromaios commented:
@elasticsearchmachine test this please

mromaios committed:
Merge branch 'main' into dense-vector-equality-support

mromaios committed:
refactor code after ExpressionEvaluator interface move

mromaios committed:
Merge branch 'main' into dense-vector-equality-support

mromaios committed:
Merge branch 'main' into dense-vector-equality-support

mromaios committed:
test: update expected value after dense_vector addition

mromaios commented:
@elasticmachine run elasticsearch-ci/pr-upgrade-part-3, elasticsearch-ci/rest-compatibility please

carlosdelest reviewed (approved):
LGTM, thank you for your work @majiayu000 and @mromaios !

mromaios commented:
@elasticmachine run Elasticsearch Serverless Checks, elasticsearch-ci/rest-compatibility please

mromaios committed:
Merge branch 'main' into dense-vector-equality-support
